Applications/Spire/Source/KeyBindings/OrderTasksTableViewModel.cpp

```diff
@@ -0,0 +1,87 @@
+#include "Spire/KeyBindings/OrderTasksTableViewModel.hpp"
+#include "Spire/Spire/ArrayListModel.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  auto adjust_row(const AnyListModel& source) {
+    auto row = std::make_shared<ArrayListModel<std::any>>();
+    static auto empty_value = std::any();
+    row->push(empty_value);
+    for(auto i = 0; i < source.get_size(); ++i) {
+      row->push(source.get(i));
+    }
+    return row;
+  }
+}
+
+OrderTasksTableViewModel::OrderTasksTableViewModel(
+  std::shared_ptr<TableModel> source)
+  : m_source(std::move(source)),
+    m_source_connection(m_source->connect_operation_signal(
+      std::bind_front(&OrderTasksTableViewModel::on_operation, this))) {}
+
+int OrderTasksTableViewModel::get_row_size() const {
+  return m_source->get_row_size() + 1;
+}
+
+int OrderTasksTableViewModel::get_column_size() const {
+  return m_source->get_column_size() + 1;
+}
+
+AnyRef OrderTasksTableViewModel::at(int row, int column) const {
+  if(row < 0 || row >= get_row_size() || column < 0 ||
+      column >= get_column_size()) {
+    throw std::out_of_range("The row or column is out of range.");
+  }
+  static auto empty_value = AnyRef();
+  if(column == 0) {
+    return empty_value;
+  }
+  if(row < m_source->get_row_size()) {
+    return m_source->at(row, column - 1);
+  }
+  return empty_value;
+}
+
+QValidator::State OrderTasksTableViewModel::set(int row, int column,
+    const std::any& value) {
+  if(row < 0 || row >= get_row_size() || column < 0 ||
```

Comment:
I don't believe these checks are needed, can you verify? Only perform checks if they are needed to satisfy the semantics of this class, there is no need to duplicate error handling that would need to be performed by the source model.

Reply:
Good points. I think checks for `row` might be needed and checks for `column` could be removed. If there are no checks for `row`, and `row` is -1 and `column` is 0, the function of set will return `Invalid`, but the function is supposed to throw an exception since the function is trying to access the row which is out of range. Validating `column` can be done by the source model.